sdk/messaging/azservicebus/internal/amqpLinks.go

```diff
@@ -97,240 +101,238 @@ type CreateLinkFunc func(ctx context.Context, session AMQPSession) (AMQPSenderCl
 		return err
 	}
 
-	links.revision++
-
 	span.AddAttributes(
 		tab.StringAttribute("outcome", "recovered"),
-		tab.StringAttribute("revision_new", fmt.Sprintf("%d", links.revision)),
+		tab.StringAttribute("revision_new", fmt.Sprintf("%d", links.id)),
 	)
 	return nil
 }
 
 // Recover will recover the links or the connection, depending
-// on the severity of the error. This function uses the `baseRetrier`
-// defined in the links struct.
-func (links *amqpLinks) RecoverIfNeeded(ctx context.Context, linksRevision uint64, origErr error) error {
+// on the severity of the error.
+func (links *AMQPLinksImpl) RecoverIfNeeded(ctx context.Context, theirID LinkID, origErr error) error {
 	ctx, span := tab.StartSpan(ctx, tracing.SpanRecover)
 	defer span.End()
 
-	var err error = origErr
-
-	retrier := links.baseRetrier.Copy()
-
-	for retrier.Try(ctx) {
-		span.AddAttributes(tab.StringAttribute("recover_attempt", fmt.Sprintf("%d", retrier.CurrentTry())))
-
-		err = links.recoverImpl(ctx, retrier.CurrentTry(), linksRevision, err)
-
-		if err == nil {
-			return nil
-		}
-	}
-
-	return err
-}
-
-func (links *amqpLinks) recoverImpl(ctx context.Context, try int, linksRevision uint64, origErr error) error {
-	_, span := tab.StartSpan(ctx, tracing.SpanRecoverLink)
-	defer span.End()
-
 	if origErr == nil || IsCancelError(origErr) {
 		return nil
 	}
 
+	log.Writef(EventConn, "Recovering link for error %s", origErr.Error())
+
 	select {
 	case <-ctx.Done():
 		return ctx.Err()
 	default:
 	}
 
-	span.AddAttributes(tab.Int64Attribute("attempt", int64(try)))
-
-	if shouldRecreateLink(origErr) {
-		span.AddAttributes(
-			tab.StringAttribute("recovery_kind", "link"),
-			tab.StringAttribute("error", origErr.Error()),
-			tab.StringAttribute("error_type", fmt.Sprintf("%T", origErr)))
+	sbe := GetSBErrInfo(origErr)
 
-		if err := links.recoverLink(ctx, &linksRevision); err != nil {
-			span.AddAttributes(tab.StringAttribute("recoveryFailure", err.Error()))
+	if sbe.RecoveryKind == RecoveryKindLink {
+		if err := links.recoverLink(ctx, theirID); err != nil {
+			log.Writef(EventConn, "failed to recreate link: %s", err.Error())
 			return err
 		}
 
+		log.Writef(EventConn, "Recovered links")
 		return nil
-	} else if shouldRecreateConnection(ctx, origErr) {
-		span.AddAttributes(
-			tab.StringAttribute("recovery_kind", "connection"),
-			tab.StringAttribute("error", origErr.Error()),
-			tab.StringAttribute("error_type", fmt.Sprintf("%T", origErr)))
-
-		if err := links.recoverConnection(ctx); err != nil {
-			span.Logger().Error(fmt.Errorf("failed to recreate connection: %w", err))
-			return err
-		}
-
-		// unconditionally recover the link if the connection died.
-		if err := links.recoverLink(ctx, nil); err != nil {
-			span.Logger().Error(fmt.Errorf("failed to recover links after connection restarted: %w", err))
+	} else if sbe.RecoveryKind == RecoveryKindConn {
+		if err := links.recoverConnection(ctx, theirID); err != nil {
```

Comment:
recoverConnection was changed so that it recovers the connection _and_ the links. Before they were separate, which meant that we weren't always recreating the links after the connection had been recreated. This was just wasteful, so now they're one operation.